src/ILVerify/src/ILImporter.StackValue.cs

```diff
@@ -92,17 +105,17 @@ static public StackValue CreateFromType(TypeDesc type)
                 case TypeFlags.UInt16:
                 case TypeFlags.Int32:
                 case TypeFlags.UInt32:
-                    return CreatePrimitive(StackValueKind.Int32);
+                    return CreatePrimitive(StackValueKind.Int32, type);
```

Comment:
That is correct.
The current implementation of ILImporter.StackValue only tracks the _intermediate type_, which is fine as long as we only use it for tracking/describing values that are stored on the evaluation stack.
I think we still need type compare methods like IsAssignable and IsCompatible to compare TypeDescs in order to be able to also compare verification types without any code duplication. In order to then call these methods with StackValues, a conversion back to TypeDesc would be needed for primitive types (which would return the intermediate type as TypeDesc instead of a StackValue), or StackValues also store the TypeDesc of the intermediate type, instead of null.
This way only the intermediate type is tracked for types on the stack, but we can still use the verification type when needed for verification of instructions.
I will implement these changes and change the StackValue tracking the exact type of primitives to tracking only the intermediate type.